File: 2262-solving-questions-with-brainpower/2262-solving-questions-with-brainpower.java
class Solution {
    public long dpf(int curr, int[][] questions, long[] dp) {
        if (curr >= questions.length) {
            return 0;
        }
        
        if(dp[curr]!=-1) return dp[curr];

        long skip = dpf(curr + 1, questions, dp);
        long take = questions[curr][0] + (curr + questions[curr][1] + 1 < questions.length 
                                          ? dpf(curr + questions[curr][1] + 1, questions, dp) 
                                          : 0);
        
        return dp[curr] = Math.max(skip, take);
    }

    public long mostPoints(int[][] questions) {
        int n = questions.length;
        long[] dp = new long[n];
        Arrays.fill(dp, -1);
        return dpf(0, questions, dp);
    }
}

File: 2262-solving-questions-with-brainpower/2262-solving-questions-with-brainpower.py
class Solution(object):
    def mostPoints(self, questions):
        """
        :type questions: List[List[int]]
        :rtype: int
        """
        n = len(questions)
        self.res = -1


        def dfs(idx, dp):
            if idx >= n:
                return 0

            if dp[idx]!=-1:
                return dp[idx]

            #Pick
            pick = questions[idx][0] + dfs(idx+questions[idx][1]+1, dp)

            #Skip
            skip = dfs(idx+1, dp)

            dp[idx] = max(pick, skip)
            return dp[idx]

        dp = [-1]*(n+1)
        dfs(0, dp)
        return max(dp)